src/content/posts/2021/pitfalls-of-unoptimzied-nuget-packages/index.md

```diff
@@ -0,0 +1,177 @@
+- src: 'unoptimized-assembly-detected.gif'
+
+---
+
+## tl;dr
+
+You might unknowingly be using an unoptimized dependency which could impact the performance of your app.
+Now, it's easy to get a warning if that happens, at build time,
+using [the NuGet package `UnoptimizedAssemblyDetector`](https://www.nuget.org/packages/UnoptimizedAssemblyDetector/):
+
+```xml
+<PackageReference Include="UnoptimizedAssemblyDetector" Version="0.0.5">
+  <PrivateAssets>all</PrivateAssets>
+</PackageReference>
+```
+
+After you add this NuGet package, a warning will be included in your build if any unoptimized assembly is detected.
+
+[![unoptimized-assembly-detected](unoptimized-assembly-detected.gif)](unoptimized-assembly-detected.gif)
+
+The source code of project is [hosted on GitHub](https://github.com/bruno-garcia/unoptimized-assembly-detector) ⭐.
+
+## What's the pitfall?
+
+The default build configuration for `dotnet pack` and `dotnet publish` is not `Release` as you might assume or expect.
+When publishing NuGet packages to _nuget.org_, one must explicitly make it a release package, like: `dotnet pack -c Release`.
+In this blog post you'll learn that's not always the case. Even for those who know, accidents happens and a misconfigured build script can result in a `Debug` build ending up on _nuget.org_.
+
+The default value being `Debug` isn't alone the reason why this is a pitfall.
+There's no warning of any kind from the point you create the NuGet package, through uploading to _nuget.org_ all the way to restoring it for your project.
+
+## Why should I care?
+
+If you need convincing, keep reading and I'll do my best.
+On the other hand if you're already convinced, feel free to skip ahead to [Sensible defaults](#sensible-defaults).
+If you don't care if the code you run was compiled with optimizations, it's unlikely this blog post is for you.
+
+### Optimizations
+
+> Never use the Debug build for benchmarking. Never. The debug version of the target method can run 10–100 times slower.
+
+Source: [Good Practices by BenchmarkDotNet](https://benchmarkdotnet.org/articles/guides/good-practices.html).
+
+The [C# compiler docs](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/code-generation#optimize) state:
+
+> The Optimize option enables or disables optimizations performed by the compiler to make your output file smaller, faster, and more efficient. The Optimize option is enabled by default for a Release build configuration. It is off by default for a Debug build configuration.
+
+If that wasn't enough to convince you, maybe Eric Lippert, former Principal Engineer in the C# language design team's blog post from 2009 will help. [Thanks to archive.org](https://web.archive.org/web/20110802071721/http://blogs.msdn.com/b/ericlippert/archive/2009/06/11/what-does-the-optimize-switch-do.aspx):
+
+* Expressions which are determined to be only useful for their side effects are turned into code that merely produces the side effects.
```

Comment:
I think I understand what this means, but I'm not sure. An example would be really nice.